cpp/tests/groupby/lists_tests.cu

```diff
@@ -64,70 +64,22 @@ static constexpr auto null = -1;
-template <typename Equal>
-struct match_expected_fn {
-  match_expected_fn(cudf::size_type const num_rows, Equal equal)
-    : _num_rows{num_rows}, _equal{equal}
-  {
-  }
-
-  __device__ bool operator()(cudf::size_type const idx)
-  {
-    for (auto i = _num_rows; i < 2 * _num_rows; i++) {
-      if (_equal(idx, i)) { return true; }
-    }
-    return false;
-  }
-
-  cudf::size_type const _num_rows;
-  Equal _equal;
-};
-
 inline void test_hash_based_sum_agg(column_view const& keys,
                                     column_view const& values,
                                     column_view const& expect_keys,
                                     column_view const& expect_vals)
 {
-  auto const include_null_keys = null_policy::INCLUDE;
-  auto const keys_are_sorted   = sorted::NO;
-
-  std::vector<groupby::aggregation_request> requests;
-  auto& request  = requests.emplace_back(groupby::aggregation_request());
-  request.values = values;
-  request.aggregations.push_back(std::move(cudf::make_sum_aggregation<groupby_aggregation>()));
-
-  groupby::groupby gb_obj(cudf::table_view({keys}), include_null_keys, keys_are_sorted);
-
-  auto result = gb_obj.aggregate(requests);
-
-  cudf::table_view result_kv{
-    {result.first->get_column(0).view(), result.second[0].results[0]->view()}};
-  cudf::table_view expected_kv{{expect_keys, expect_vals}};
-
-  auto const num_rows = result_kv.num_rows();
-  EXPECT_EQ(num_rows, expected_kv.num_rows());
-
-  // Concatenate expected table and resulting table into one unique table `t`:
-  // expected table:  `t [       0,     num_rows - 1]`
-  // resulting table: `t [num_rows, 2 * num_rows - 1]`
-  auto combined_table = cudf::concatenate(std::vector{expected_kv, result_kv});
-  auto preprocessed_t = cudf::experimental::row::hash::preprocessed_table::create(
-    combined_table->view(), cudf::get_default_stream());
-  cudf::experimental::row::equality::self_comparator comparator(preprocessed_t);
-
-  auto const null_keys_are_equal =
-    include_null_keys == null_policy::INCLUDE ? null_equality::EQUAL : null_equality::UNEQUAL;
-  auto row_equal = comparator.equal_to(nullate::DYNAMIC{true}, null_keys_are_equal);
-  auto func      = match_expected_fn{num_rows, row_equal};
-
-  // For each row in expected table `t[0, num_rows)`, there must be a match
-  // in the resulting table `t[num_rows, 2 * num_rows)`
-  EXPECT_TRUE(thrust::all_of(rmm::exec_policy(cudf::get_default_stream()),
-                             thrust::make_counting_iterator<cudf::size_type>(0),
-                             thrust::make_counting_iterator<cudf::size_type>(num_rows),
-                             func));
+  auto const sort_order         = sorted_order(table_view{{expect_keys}}, {}, {null_order::AFTER});
+  auto const sorted_expect_keys = gather(table_view{{expect_keys}}, *sort_order);
+  auto const sorted_expect_vals = gather(table_view{{expect_vals}}, *sort_order);
```

Comment:
Can you sort expected keys/values in `test_single_agg` instead of `test_hash_based_sum_agg`? References need to be sorted for sort-based implementations as well.

Reply:
Yeah. So we sort everything in the `test_single_agg` function and can use it everywhere. And `test_hash_based_sum_agg` can be removed completely.

Reply:
Sure, I can. But can you clear what this https://github.com/divyegala/cudf/blob/732baf856dd88d8ae548be9fdbd883f0cc1f8ade/cpp/tests/groupby/lists_tests.cu#L99 function is doing? I don't understand why the tests first do a sort-based-groupby and then a hash-based-groupby

Reply:
Groupby has 2 approaches: sort-based implementation and hash-based implementation. That function is testing both approaches (same input, same output, different way to compute).

I've recommended to remove all `test_sort_based_sum_agg`+`test_hash_based_sum_agg` and just call `test_single_agg(force_use_sort_impl::NO)` directly. But I changed my mind, as keeping them makes the code looks more organized so just ignore my last comment above.

Reply:
Looks like a lot of other tests start failing if I sort expected keys/vals in `test_single_agg` @PointKernel. I pushed the latest and the changes look simple enough to me so not sure why tests are failing

Reply:
You sort the expected keys/vals but I don't see you sort the result keys/vals?

Reply:
That's part of the implementation, right? sort-based sorts the results and hash-bashed sorts it explicitly in the test if you scroll below to the if-else block

Reply:
Yes. But the way of sorting in groupby is parameterized---nulls can be before or after non-nulls depending on the parameter. In `test_single_agg` it is fixed as `AFTER`, so if the result is sorted by `BEFORE` you will have mismatched output.

Reply:
If you want to avoid sorting twice, you may need to investigate the input parameter `null_precedence` of that function and sort the expected keys/vals accordingly.
If after doing so you still have test failed then we can investigate further to see what's wrong.